LIBITD-2546. Integrated models into API views

neoladsouza opened the pull request:
Removed placeholder data and queried the database models for each view. Modified each API view's tests to mock data from the database.

https://umd-dit.atlassian.net/browse/LIBITD-2546

Commit:
LIBITD-2546. Integrated models into API views

Removed placeholder data and queried the database models
for each view. Modified each API view's tests to mock
data from the database.

https://umd-dit.atlassian.net/browse/LIBITD-2546

peichman-umd commented on src/ipmanager/api/views.py:
```diff
@@ -1,21 +1,22 @@
-                "@id": request_url + group_key,
-                "key": group_key,
-                "name": group_key + "'s name"
+                "@id": request.build_absolute_uri(reverse("group_key", args=[group.key])),
```
Since you build this same structure three times, I would suggest creating a function for it:

```python
def build_group_metadata(request, group):
    return {
        "@id": request.build_absolute_uri(reverse("group_key", args=[group.key])),
        "key": group.key,
        "name": group.name,
    }
```
Then you can use it in this function, as well as later on lines 81 and 96.

peichman-umd commented on src/ipmanager/api/views.py:
```diff
@@ -1,21 +1,22 @@
+                "name": group.name,
             })
 
         body = {
```
If you encapsulate the building group metadata dictionary as described above, you can simplify the construction of the groups list into a list comprehension:

```python
groups = [build_group_metadata(group) for group in exported_groups]
```

peichman-umd commented on src/ipmanager/api/views.py:
```diff
@@ -24,37 +25,36 @@ def get(self, request):
             }
-
+            
             return JsonResponse(body)
```
Here you can inline the call to `build_group_metadata()`, and remove the `body` variable altogether:

```python
return JsonResponse(build_group_metadata(request, group))
```

peichman-umd commented on src/ipmanager/api/views.py:
```diff
@@ -72,29 +72,34 @@ def get(self, request):
+def generateChecksArray(ip_address, all_groups, request):
+    result = []
+    for group in all_groups:
+        single_group = {}
```
A better pattern here is to use a single dictionary literal to create the `single_group` dictionary, instead of defining it by each key separately:

```python
single_group = {
    "@id": request.build_absolute_uri(reverse("check")) + "?" + urlencode({"group":group.key, "ip":ip_address}),
    "group": build_group_metadata(request, group),
    "ip": ip_address,
    "contained": contained(ip_address, group.key),
}
```

This also uses the `build_group_metadata()` function described earlier, and fixes the key names (`"@group"` --> `"group"` and `"ip_address"` --> `"ip"`).

peichman-umd commented on src/ipmanager/api/views.py:
```diff
@@ -72,29 +72,34 @@ def get(self, request):
-# For contained, we need to create a function to actually check for membership and return boolean value.
 
-def contained():
+def generateChecksArray(ip_address, all_groups, request):
```
I would suggest renaming this function to `build_check_result`, and changing it to operate on a single group. That way, you can reuse it where you are only checking a single group (line 80 above), as well as here when you are checking multiple groups (line 75). The multiple groups case is another good candidate for a list comprehension:

```python
{
    ...
    "checks": [build_check_result(ip_address, group, request) for group in all_groups],
    ...
}
```

peichman-umd reviewed (changes requested):
The logic is good. I have added several inline comments highlighting where I think encapsulating some small bots of functionality into functions will lead to clearer code, as well as making it easier to maintain by reducing duplication.

Commit:
LIBITD-2546. Created functions for building group and checks metadata

Implemented changes to the views.py file by creating new functions
and using them in the views.

https://umd-dit.atlassian.net/browse/LIBITD-2546
